source/agora/node/Registry.d

```diff
@@ -354,18 +354,15 @@ public class NameRegistry: NameRegistryAPI
                 break;
             }
 
-            if (this.validators.owns(q.qname))
-                this.validators.answer(q, reply);
-            else if (this.flash.owns(q.qname))
-                this.flash.answer(q, reply);
-            else
-            {
-                log.warn("Refusing {} for unknown zone: {}", q.qtype, q.qname);
-                reply.header.RCODE = Header.RCode.Refused;
-            }
+            reply.header.RCODE = this.validators.answer(q, reply);
+            if (reply.header.RCODE != Header.RCode.NoError)
+                reply.header.RCODE = this.flash.answer(q, reply);
 
             if (reply.header.RCODE != Header.RCode.NoError)
+            {
+                log.warn("Refusing {} for unknown zone: {}", q.qtype, q.qname);
                 break;
+            }
```

Comment:
I think it's not sound to depend so much on the internals here. The top level should match the `ZoneData` instead of cascading through them. It would avoid falling back to `flash` when a request can be fulfilled (e.g. a query for an inexistent validator will still go through the `flash` `ZoneData`.

Reply:
What do you think about checking for `Refused` instead of `NoError`?
Like;
`if (reply.header.RCODE != Header.RCode.Refused)`

Reply:
It has the same problem. For example, we could implement whitelisting for AXFR later, and we would then return `Refused` for queries for our zone that do not match the whitelist.